kuksa_apps/kanto/edge_client.py

```diff
@@ -0,0 +1,122 @@
+
+    def on_message(self, client, userdata, msg):
+        try:
+            if msg.topic == EDGE_CLOUD_CONNECTOR_TOPIC_DEV_INFO:
+                if self.device_info is None:
+                    self.device_info = EdgeDeviceInfo()
+                    self.device_info.unmarshal_json(msg.payload)
+                    self.add_vss_feature()
+                    self.subscribe()
+                else:
+                    print('device info already available - discarding message')
+                return
+        except Exception as ex:
+            print(ex)
+
+    def subscribe(self):
+        self.kuksa_client.subscribe('Vehicle.Cabin.Infotainment.Navigation.CurrentLocation.Altitude', self.on_kuksa_signal)
+        self.kuksa_client.subscribe('Vehicle.Cabin.Infotainment.Navigation.CurrentLocation.Latitude', self.on_kuksa_signal)
+        self.kuksa_client.subscribe('Vehicle.Cabin.Infotainment.Navigation.CurrentLocation.Longitude', self.on_kuksa_signal)
+        # self.kuksa_client.subscribe('Vehicle.Cabin.Infotainment.Navigation.CurrentLocation.Accuracy', self.on_kuksa_signal)
+        self.kuksa_client.subscribe('Vehicle.Cabin.Infotainment.Navigation.CurrentLocation.Speed', self.on_kuksa_signal)
+
+    def add_vss_feature(self):
+        # add the vss feature
+        cmd = Command(self.device_info.deviceId).feature(FEATURE_ID_VSS).modify(Feature().to_ditto_dict())
+        cmd_envelope = cmd.envelope(response_required=False, content_type="application/json")
+        self.ditto_client.send(cmd_envelope)
+
+        # add the vss tree as properties
+        vss_tree = self.kuksa_client.getValue('Vehicle/Cabin/Infotainment/Navigation/CurrentLocation/*')
+        processed = process_tree(vss_tree)
+        for key, val in processed.items():
+            if 'Heading' in key or 'Accuracy' in key:
+                continue
+            cmd = Command(self.device_info.deviceId).feature_property(FEATURE_ID_VSS, key).modify(val)
+            cmd_envelope = cmd.envelope(response_required=False, content_type="application/json")
+            self.ditto_client.send(cmd_envelope)
+
+    def on_kuksa_signal(self, message):
+        if self.device_info is None:
+            print('no device info is initialized to process VSS data for')
+            return
+        path, val = process_signal(message)
+        # update property
+        cmd = Command(self.device_info.deviceId).feature_property(FEATURE_ID_VSS, path.replace('.','/')).modify(val)
+        cmd_envelope = cmd.envelope(response_required=False, content_type="application/json")
+        self.ditto_client.send(cmd_envelope)
+
+    def shutdown(self):
+        self.kuksa_client.stop()
+        self.ditto_client.disconnect()
+
+
+if __name__ == "__main__":
+    paho_client = mqtt.Client()
+    edge_client = EdgeClient()
+
+    paho_client.on_connect = edge_client.on_connect
+    paho_client.on_message = edge_client.on_message
+    paho_client.connect("localhost")
```

Comment:
The configuration from the .ini should be checked and applied here if provided.

Reply:
I pulled the value out of the config.ini and backed it with a localhost fallback.